Allow output of ES2015-style Unicode regexes

ticky opened the pull request:
This is a polished up version of what I was working on as part of #15; it updates regexgen to optionally output ES6-compatible Unicode regexes.

It does this by looking at the flags passed; if `u` is included, it outputs a Unicode regex rather than an UCS-2 style regex, by passing the requisite options down to `regenerate` and `jsesc`.

~This isn’t the _cleanest_ change, as no state was previously passed down into the AST methods, so it required quite a bit of passing around. I also only added the additional arguments to the AST object classes affected by the change!~ State is no longer passed down into AST constructors as of 7b560c7, instead being passed to `toString` methods to control their behaviour directly!

fixes #15

Todo:
- [x] Clean up for PR
- [x] Write specs for the updated behaviour
- [x] Document updated behaviour

ticky commented:
I’ve had some trouble with one of the existing tests - I must’ve missed something here;

```
  17 passing (43ms)
  1 failing

  1) regexgen should escape meta characters:

      AssertionError [ERR_ASSERTION]: /foo|bar[test]+/ deepEqual /foo\|bar\[test\]\+/
      + expected - actual

      -/foo|bar[test]+/
      +/foo\|bar\[test\]\+/

      at Context.<anonymous> (test/test.js:34:12)
```

I don’t _think_ anything I did should’ve changed such escaping behaviour. I couldn’t find the source of the bug!

sj26 commented on src/ast.js:
```diff
@@ -130,12 +136,16 @@ class Literal {
   }
 
   toString() {
-    return jsesc(this.value).replace(/([\t\n\f\r\$\(\)\*\+\-\.\?\[\]\^\{\|\}])/g, '\\$1');
```
@ticky I think maybe you're removed escaping here which is causing the failure?

ticky replied:
Oh dang I totally missed that in my readings, yeah!

ticky commented:
Thanks to @sj26’s sharp eye, I’ve updated this and it now passes all the original specs, and I’ve added a couple of extras 😃🎉

devongovett commented on src/ast.js:
```diff
@@ -130,12 +136,32 @@ class Literal {
+
+        // don't escape closing curly braces ending Unicode escapes
+        if (match === '}' && string.slice(offset - 8, offset - 5) === '\\u{') {
+          return match;
```
I think this won't always work since unicode escapes can be variable length, e.g. `\u{10ffff}`. Maybe try something like this:

```javascript
.replace(/(\\u\{[a-z1-9]+\})|([\{\}])/ig, (m, u, e) => u || '\\' + e)
```

ticky replied:
Oh, good point!

devongovett commented on src/regex.js:
```diff
@@ -20,30 +21,30 @@ function toRegex(root) {
+        B[i] = union(B[i], concat(A[i][n], B[n], flags), flags);
         for (let j = 0; j < n; j++) {
-          A[i][j] = union(A[i][j], concat(A[i][n], A[n][j]));
+          A[i][j] = union(A[i][j], concat(A[i][n], A[n][j], flags), flags);
```
What if we just passed the flags to `toString` instead of the constructor of every AST node? Seems like it would simplify this.

ticky replied:
Hmm, will that then pass it down into other AST nodes’ `toString`s? I found the code around that quite hard to follow.

ticky replied:
Got it working! :)

devongovett commented on src/ast.js:
```diff
@@ -30,26 +30,31 @@ class Alternation {
   }
 
+  get isSingleCharacter() {
+    return !this.set.toArray().some(c => c > 0xffff);
```
This should probably always return true if in unicode mode. Will get rid of the parentheses around the second character in your test case, which aren't necessary:

`\u261D(?:\u{1F3FB})?/u`

devongovett replied:
You may also need to modify `Literal#isSingleCharacter` as well to count code points instead of code units [here](https://github.com/devongovett/regexgen/blob/a5683bc6d4a2d84d76bfa792f967ec9852c31090/src/ast.js#L130). You can do that with `Array.from(this.value).length`.

ticky replied:
Good thought, I had meant to ask about this one! 😃

ticky replied:
I ended up adding an `isSingleCodepoint` getter to these two, since the context of the flags has been moved to only be in `toString` and, in this case, `parens` - that way the unicode context doesn’t need to be passed to these particular methods.

ticky commented:
Thanks for your constructive input on this, @devongovett 😃

ticky commented:
As part of trying to integrate the new behaviour in mathiasbynens/emoji-regex, I found a little bug, which I fixed in b737de8!

ticky commented:
I think I lack compatibility with the changes from #16. Oops!

devongovett commented on test/test.js:
```diff
@@ -69,6 +70,14 @@ describe('regexgen', function () {
+      /\u{1F3F4}(?:\u{E0067}\u{E0062}(?:\u{E0073}\u{E0063}\u{E0074}|\u{E0077}\u{E006C}\u{E0073}|\u{E0065}\u{E006E}\u{E0067}))?/u
+    );
+  });
+
```
Can you also add a test case for unicode character classes, i.e. a range?

devongovett commented on src/ast.js:
```diff
@@ -39,16 +39,22 @@ class CharClass {
-  getCharClass() {
-    return this.set;
+  get isSingleCodepoint() {
+    return Array.from(this.set).length === 1;
```
In unicode mode, character classes can always be treated as a single "codepoint" because that's what they match. In ucs2 mode, regenerate might have to return a sequence instead, which is why we check if any of the values are > 0xffff.

ticky commented:
I’ve added the spec and change you requested, and also added a spec for the failure I was seeing upstream in mathiasbynens/emoji-regex#22, which now appears to be corrected 😃

Thanks again for your patience and your fantastic help!

devongovett commented:
@ticky thanks so much, awesome work on this! Released in v1.3.0. 🎉
